extensions/arc/notebooks/arc/content/postgres/tsg100-troubleshoot-postgres.ipynb

```diff
@@ -0,0 +1,590 @@
+    "\n",
+    "        raise FileNotFoundError(f\"Executable '{cmd_actual[0]}' not found in path (where/which)\")\n",
+    "    else:   \n",
+    "        cmd_actual[0] = which_binary\n",
+    "\n",
+    "    start_time = datetime.datetime.now().replace(microsecond=0)\n",
+    "\n",
+    "    print(f\"START: {cmd} @ {start_time} ({datetime.datetime.utcnow().replace(microsecond=0)} UTC)\")\n",
+    "    print(f\"       using: {which_binary} ({platform.system()} {platform.release()} on {platform.machine()})\")\n",
+    "    print(f\"       cwd: {os.getcwd()}\")\n",
+    "\n",
+    "    # Command-line tools such as CURL and AZDATA HDFS commands output\n",
+    "    # scrolling progress bars, which causes Jupyter to hang forever, to\n",
+    "    # workaround this, use no_output=True\n",
+    "    #\n",
+    "\n",
+    "    # Work around a infinite hang when a notebook generates a non-zero return code, break out, and do not wait\n",
+    "    #\n",
+    "    wait = True \n",
+    "\n",
+    "    try:\n",
+    "        if no_output:\n",
+    "            p = Popen(cmd_actual)\n",
+    "        else:\n",
+    "            p = Popen(cmd_actual, stdout=PIPE, stderr=PIPE, bufsize=1)\n",
+    "            with p.stdout:\n",
+    "                for line in iter(p.stdout.readline, b''):\n",
+    "                    line = line.decode()\n",
+    "                    if return_output:\n",
+    "                        output = output + line\n",
+    "                    else:\n",
+    "                        if cmd.startswith(\"azdata notebook run\"): # Hyperlink the .ipynb file\n",
+    "                            regex = re.compile('  \"(.*)\"\\: \"(.*)\"') \n",
+    "                            match = regex.match(line)\n",
+    "                            if match:\n",
+    "                                if match.group(1).find(\"HTML\") != -1:\n",
+    "                                    display(Markdown(f' - \"{match.group(1)}\": \"{match.group(2)}\"'))\n",
+    "                                else:\n",
+    "                                    display(Markdown(f' - \"{match.group(1)}\": \"[{match.group(2)}]({match.group(2)})\"'))\n",
+    "\n",
+    "                                    wait = False\n",
+    "                                    break # otherwise infinite hang, have not worked out why yet.\n",
+    "                        else:\n",
+    "                            print(line, end='')\n",
+    "                            if rules is not None:\n",
+    "                                apply_expert_rules(line)\n",
+    "\n",
+    "        if wait:\n",
+    "            p.wait()\n",
+    "    except FileNotFoundError as e:\n",
+    "        if install_hint is not None:\n",
+    "            display(Markdown(f'HINT: Use {install_hint} to resolve this issue.'))\n",
+    "\n",
+    "        raise FileNotFoundError(f\"Executable '{cmd_actual[0]}' not found in path (where/which)\") from e\n",
```

Comment:
here too

Reply:
See other reply.  I assume they refer to the where/which programs that look up executable locations, but I'm just referencing an existing snippet here and was not the author.